src/request/sign-swap/SignSwapApi.js

```diff
@@ -145,14 +153,37 @@ class SignSwapApi extends PolygonRequestParserMixin(BitcoinRequestParserMixin(To
                 amount: this.parsePositiveInteger(request.redeem.amount, false, 'redeem.amount'),
             };
         } else if (request.redeem.type === 'EUR') {
+            const settlement = this.parseOasisSettlementInstruction(request.redeem.settlement, 'redeem.settlement');
+            if (
+                settlement.type !== 'sepa'
+                && (CONFIG.NETWORK === Constants.NETWORK.MAIN || settlement.type !== 'mock')
+            ) {
+                throw new Errors.InvalidRequestError('Invalid redeeming settlement type');
+            }
             parsedRequest.redeem = {
                 type: 'EUR',
                 keyPath: this.parsePath(request.redeem.keyPath, 'redeem.keyPath'),
-                settlement: this.parseOasisSettlementInstruction(request.redeem.settlement, 'redeem.settlement'),
+                settlement,
                 amount: this.parsePositiveInteger(request.redeem.amount, false, 'redeem.amount'),
                 fee: this.parsePositiveInteger(request.redeem.fee, true, 'redeem.fee'),
                 bankLabel: this.parseLabel(request.redeem.bankLabel, true, 'redeem.bankLabel'),
             };
+        } else if (request.redeem.type === 'CRC') {
+            const settlement = this.parseOasisSettlementInstruction(request.redeem.settlement, 'redeem.settlement');
+            if (
+                settlement.type !== 'sinpemovil'
+                && (CONFIG.NETWORK === Constants.NETWORK.MAIN || settlement.type !== 'mock')
+            ) {
+                throw new Errors.InvalidRequestError('Invalid redeeming settlement type');
+            }
+            parsedRequest.redeem = {
+                type: 'CRC',
+                keyPath: this.parsePath(request.redeem.keyPath, 'redeem.keyPath'),
+                settlement,
+                amount: this.parsePositiveInteger(request.redeem.amount, false, 'redeem.amount'),
+                fee: this.parsePositiveInteger(request.redeem.fee, true, 'redeem.fee'),
+                recipientLabel: this.parseLabel(request.redeem.recipientLabel, true, 'redeem.recipientLabel'),
+            };
         } else {
             throw new Errors.InvalidRequestError('Invalid redeeming type');
         }
```

Comment:
There is the assumption throughout the code that only one side of the swap, funding or redeeming, is a fiat currency. Thus a check should be added to the parser, that not both sides are a fiat currency.

Additionally, but not directly related, a check should be added, that the asset on both sides are not the same. Such check has been missing so far, from what it seems.

Reply:
I think we do have a check if both sides are the same:

https://github.com/nimiq/keyguard/blob/b6ebfd4953ac82aaf0f618df186db2439685ef07/src/request/sign-swap/SignSwapApi.js#L34-L36

I also added a check to avoid fiat-to-fiat swaps in eefcb03075bc444c5405243cf6fb00f29dfb2eb3